### ISSUE TITLE
May be uninitialized warnings

### ISSUE THREAD

**Issue by Kojoley** [boostorg/function#27]
```
gcc.compile.c++ bin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-03-iso\visibility-hidden\function_n_test.o
In file included from ./boost/function/detail/maybe_include.hpp:15,
                 from ./boost/function/function0.hpp:11,
                 from ./boost/test/execution_monitor.hpp:28,
                 from ./boost/test/impl/execution_monitor.ipp:28,
                 from ./boost/test/minimal.hpp:53,
                 from libs\function\test\function_n_test.cpp:10:
./boost/function/function_template.hpp: In function 'void test_construct_destroy_count()':
./boost/function/function_template.hpp:998:24: error: '*((void*)& f +8)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
             std::memcpy(this->functor.data, f.functor.data, sizeof(this->functor.data));
             ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus.exe: all warnings being treated as errors

    "g++"   -std=c++03 -fvisibility-inlines-hidden -m64 -O3 -finline-functions -Wno-inline -Wall -Werror -fvisibility=hidden  -DBOOST_ALL_NO_LIB=1 -DNDEBUG  -I"." -c -o "bin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-03-iso\visibility-hidden\function_n_test.o" "libs\function\test\function_n_test.cpp"

...failed gcc.compile.c++ bin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-03-iso\visibility-hidden\function_n_test.o...
...skipped <pbin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-03-iso\visibility-hidden>function_n_test.exe for lack of <pbin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-03-iso\visibility-hidden>function_n_test.o...
...skipped <pbin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-03-iso\visibility-hidden>function_n_test.run for lack of <pbin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-03-iso\visibility-hidden>function_n_test.exe...
gcc.compile.c++ bin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-11-iso\visibility-hidden\function_n_test.o
In file included from ./boost/function/detail/maybe_include.hpp:15,
                 from ./boost/function/function0.hpp:11,
                 from ./boost/test/execution_monitor.hpp:28,
                 from ./boost/test/impl/execution_monitor.ipp:28,
                 from ./boost/test/minimal.hpp:53,
                 from libs\function\test\function_n_test.cpp:10:
./boost/function/function_template.hpp: In function 'void test_construct_destroy_count()':
./boost/function/function_template.hpp:998:24: error: '*((void*)& f +8)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
             std::memcpy(this->functor.data, f.functor.data, sizeof(this->functor.data));
             ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus.exe: all warnings being treated as errors

    "g++"   -std=c++11 -fvisibility-inlines-hidden -m64 -O3 -finline-functions -Wno-inline -Wall -Werror -fvisibility=hidden  -DBOOST_ALL_NO_LIB=1 -DNDEBUG  -I"." -c -o "bin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-11-iso\visibility-hidden\function_n_test.o" "libs\function\test\function_n_test.cpp"

...failed gcc.compile.c++ bin.v2\libs\function\test\function_n_test.test\gcc-8.2.0\release\cxxstd-11-iso\visibility-hidden\function_n_test.o...
```

**Comment by pdimov** [comment]
Technically correct, but I'm not sure what we can do about these. Value-initializing `functor` in the constructor will be too expensive.

**Comment by jeking3** [comment]
`-fvisibility-inlines-hidden` is not a common setting to use, right?

**Comment by jeking3** [comment]
Does it happen with `cxxstd=11` or later?

**Comment by Kojoley** [comment]
> `-fvisibility-inlines-hidden` is not a common setting to use, right?

It is default flags for `./b2 -j3 toolset=gcc-8.2.0 cxxstd=03,11 libs/function/test warnings=on warnings-as-errors=on variant=debug,release`

> Does it happen with `cxxstd=11` or later?

Yes, it is shown in the first post.

**Comment by pdimov** [comment]
Should be fixed by #38.
